source/common/router/router.cc

```diff
@@ -715,14 +716,17 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
         continue;
       }
       auto shadow_headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>(*shadow_headers_);
-      auto options = Http::AsyncClient::RequestOptions()
-                         .setTimeout(timeout_.global_timeout_)
-                         .setParentSpan(callbacks_->activeSpan())
-                         .setChildSpanName("mirror")
-                         .setSampled(shadow_policy.traceSampled())
-                         .setIsShadow(true)
-                         .setBufferAccount(callbacks_->account())
-                         .setBufferLimit(callbacks_->decoderBufferLimit());
+      auto options =
+          Http::AsyncClient::RequestOptions()
+              .setTimeout(timeout_.global_timeout_)
+              .setParentSpan(callbacks_->activeSpan())
+              .setChildSpanName("mirror")
+              .setSampled(shadow_policy.traceSampled())
+              .setIsShadow(true)
+              .setBufferAccount(callbacks_->account())
+              // A buffer limit of 1 is set in the case that retry_shadow_buffer_limit_ == 0,
+              // because a buffer limit of zero on async clients is interpreted as no buffer limit.
+              .setBufferLimit(1 > retry_shadow_buffer_limit_ ? 1 : retry_shadow_buffer_limit_);
```

Comment:
Would the `retry_shadow_buffer_limit_ == 0` case result in shadowing being effectively disabled unless the body happens to fit in a 1 byte buffer? 

If there's no way to disable buffering altogether for the async client maybe add a TODO about that, setting it to 1 seems like a little bit of a hack (but okay for now)